### DISCUSSION TITLE
GraphQL ?

### DISCUSSION THREAD

**Discussion by ZibanPirate** [dzcode-io/dzcode.io#445] [Architecture]
GraphQL is a nice to have skill for both front and back end developers, should we add it to dzcode.io and learn about it while working on it?

Any suggestions on what library to use, on both frontend and backend, what is the best way to use this in a monorepo? how can we take advantage of the existing Classes and Types in `@dzcode.io/models`?

**Comment by ZibanPirate** [comment]
for Frontend:
We can use Facebook's [relay](https://relay.dev/), it may require some [specific configurations](https://relay.dev/docs/guides/graphql-server-specification/) for the GraphQL server, but it's worth trying out, as it's battle-tested and just like React, it is also backed by Facebook.

**Reply by omdxp** [reply]
It's worth trying, I used [Apollo](https://www.apollographql.com) before. Neither of them is considered to be better

**Reply by ZibanPirate** [reply]
And what about state management, we currently use Redux, should we replace it completely with the GraphQL client library? Or should we keep using Redux for non-fetched data, such as Settings, global loaders ...etc

**Reply by omdxp** [reply]
I think we should keep it for non-fetched data

**Comment by ZibanPirate** [comment]
for Backend:

Since we don't have a database, it will be easier to set up a GraphQL server, and create resolvers for each module in the `./api/[module]/resolver.ts`, and since we use Dependency Injection here, we can use [TypeGraphQL](https://typegraphql.com/) and take advantage of the Type-friendly setup we have.

**Reply by omdxp** [reply]
How about we implement the GraphQL as a middleware between the database and request? meaning we can have a Postgres Database and and we query the data using GraphQL

I think it'll be good for us to implement Postgres as well and work with Docker images and create Docker network from now on, what do you think?

**Reply by ZibanPirate** [reply]
Having a database is not needed for now, we get our data from GitHub (and Gitlab/Bitbucket when needed), if we have the need for a Database, we can discuss it then.

**Reply by omdxp** [reply]
@ZibanPirate what about using [yoga](https://github.com/dotansimha/graphql-yoga) as a GraphQL server, I've just used it and I think it's pretty simple to setup especially with Prisma and pothos

**Comment by abderrahmaneMustapha** [comment]
Hi @ZibanPirate  and @Omar-Belghaouti  ,  Graphql is a good  i would love to help

**Comment by omdxp** [comment]
Hey @ZibanPirate , I found what I think it's better replacement than GraphQL, check this [tRPC](https://github.com/trpc/trpc) out! It is definitely well suited for monorepos same as ours. Basically with it, you can define methods in the api folder for queries and mutations and you will get type inference between clients (`web` and `mobile`) and the api.

**Reply by ZibanPirate** [reply]
interesting, we currently have something simular in 
https://github.com/dzcode-io/dzcode.io/blob/5e548baec797104fc9607f4ac8e1fe031096136b/api/src/app/endpoints.ts#L10-L14

tRPC is a bit more automated, but then again we have to build our routes using `@trpc/server`, this would be tricky with the way we do our routing using `routing-controllers` (which provide types, as well as validation and OpenAPI docs and a nice way of inverting control (IoC))

**Reply by omdxp** [reply]
Yeah, it takes time to reimplement all things in tRPC at first, but then again it'll be so easy to make calls from client to server as all things are inferred like the methods and request bodies.

Basically we're having the same thing with `Endpoints` interface.
